File: go/api/options/hscan_options.go
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

package options

// This struct represents the optional arguments for the HSCAN command.
type HashScanOptions struct {
	BaseScanOptions
	noValue bool
}

func NewHashScanOptionsBuilder() *HashScanOptions {
	return &HashScanOptions{}
}

/*
If this value is set to true, the HSCAN command will be called with NOVALUES option.
In the NOVALUES option, values are not included in the response.
*/
func (hashScanOptions *HashScanOptions) SetNoValue(noValue bool) *HashScanOptions {
	hashScanOptions.noValue = noValue
	return hashScanOptions
}

func (hashScanOptions *HashScanOptions) SetMatch(match string) *HashScanOptions {
	hashScanOptions.BaseScanOptions.SetMatch(match)
	return hashScanOptions
}

func (hashScanOptions *HashScanOptions) SetCount(count int64) *HashScanOptions {
	hashScanOptions.BaseScanOptions.SetCount(count)
	return hashScanOptions
}

func (options *HashScanOptions) ToArgs() ([]string, error) {
	args := []string{}
	baseArgs, err := options.BaseScanOptions.ToArgs()
	args = append(args, baseArgs...)

	if options.noValue {
		args = append(args, NoValue)
	}
	return args, err
}

File: JNI-java/client/src/main/java/glide/api/models/commands/scan/HScanOptions.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide.api.models.commands.scan;

import glide.api.commands.HashBaseCommands;
import glide.api.models.GlideString;
import glide.utils.ArgsBuilder;
import java.util.Arrays;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.experimental.SuperBuilder;

/**
 * Optional arguments for {@link HashBaseCommands#hscan(String, String, HScanOptions)}.
 *
 * @see <a href="https://valkey.io/commands/hscan/">valkey.io</a>
 */
@SuperBuilder
@EqualsAndHashCode(callSuper = false)
public class HScanOptions extends BaseScanOptions {

    /** Option string to include in the HSCAN command when values are not included. */
    public static final String NO_VALUES_API = "NOVALUES";

    /**
     * When set to true, the command will not include values in the results. This option is available
     * from Valkey version 8.0.0 and above.
     */
    @Builder.Default protected boolean noValues = false;

    @Override
    public String[] toArgs() {
        return Arrays.stream(toGlideStringArgs()).map(GlideString::getString).toArray(String[]::new);
    }

    /**
     * Creates the arguments to be used in <code>HSCAN</code> commands.
     *
     * @return a GlideString array that holds the options and their arguments.
     */
    @Override
    public GlideString[] toGlideStringArgs() {
        ArgsBuilder builder = new ArgsBuilder();

        // Add options from the superclass
        GlideString[] superArgs = super.toGlideStringArgs();
        for (GlideString arg : superArgs) {
            builder.add(arg);
        }

        // Add the noValues option if applicable
        if (noValues) {
            builder.add(NO_VALUES_API);
        }

        return builder.toArray();
    }
}
